### PR TITLE
Handle misencoding of login_source cfg in mssql

### LINKED ISSUE #16252: Change LDAP - PANIC: runtime error: invalid memory address or nil pointer dereference
- Gitea version (or commit ref): v1.14.3 or v1.13.7

- Git version: 2.32.0

- Operating system: Windows Server Core 2019
  - Git-go Version
  - gitea-1.14.3-windows-4.0-amd64
  - Docker and / or local 

- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [X] MSSQL
  - [ ] SQLite

- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [X] No

- Log gist (local):

2021/06/25 20:23:10 ...l/manager_windows.go:85:start() [T] Not running a service ... using the debug SVC manager
2021/06/25 20:23:10 ...l/manager_windows.go:101:Execute() [T] Awaiting server start-up
2021/06/25 20:23:10 cmd/web.go:82:runWeb() [I] Starting Gitea on PID: 5508
2021/06/25 20:23:10 ...dules/setting/git.go:101:newGit() [I] Git Version: 2.32.0, Wire Protocol Version 2 Enabled
2021/06/25 20:23:10 cmd/web.go:126:runWeb() [I] Global init
2021/06/25 20:23:10 ...dules/setting/git.go:101:newGit() [I] Git Version: 2.32.0, Wire Protocol Version 2 Enabled
2021/06/25 20:23:11 routers/init.go:134:GlobalInit() [T] AppPath: G:/Gitea/gitea-amd64.exe
2021/06/25 20:23:11 routers/init.go:135:GlobalInit() [T] AppWorkPath: G:/Gitea
2021/06/25 20:23:11 routers/init.go:136:GlobalInit() [T] Custom path: G:/Gitea/custom
2021/06/25 20:23:11 routers/init.go:137:GlobalInit() [T] Log path: G:/data/log
2021/06/25 20:23:11 routers/init.go:49:checkRunMode() [I] Run Mode: Prod
2021/06/25 20:23:12 ...dules/setting/log.go:287:newLogService() [I] Gitea v1.14.3 built with GNU Make 4.1, go1.16.5 : bindata, sqlite, sqlite_unlock_notify
2021/06/25 20:25:46 ...uters/routes/base.go:153:1() [E] PANIC: runtime error: invalid memory address or nil pointer dereference
        /usr/local/go/src/runtime/panic.go:212 (0x1003864)
        /usr/local/go/src/runtime/signal_windows.go:239 (0x100371e)
        /source/models/login_source.go:262 (0x2c7981a)
        /source/routers/admin/auths.go:309 (0x2c7982b)
        /source/modules/web/route.go:64 (0x2ba1ffb)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/vendor/github.com/go-chi/chi/mux.go:436 (0x274ba6a)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/modules/web/route.go:103 (0x2ba26e9)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/modules/web/route.go:103 (0x2ba26e9)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/modules/web/route.go:103 (0x2ba26e9)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/modules/web/route.go:103 (0x2ba26e9)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/vendor/github.com/go-chi/chi/middleware/get_head.go:37 (0x2dd3fe1)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/modules/context/context.go:704 (0x276b121)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/routers/routes/base.go:94 (0x2ddb6ac)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/routers/routes/base.go:94 (0x2ddb6ac)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/modules/public/public.go:86 (0x1fd800c)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/modules/public/public.go:86 (0x1fd800c)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/routers/routes/base.go:199 (0x2ddd317)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/vendor/gitea.com/go-chi/session/session.go:256 (0x212ed6e)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/vendor/github.com/go-chi/chi/mux.go:70 (0x274956a)
        /source/vendor/github.com/go-chi/chi/mux.go:311 (0x274fa1b)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/vendor/github.com/go-chi/chi/mux.go:436 (0x274ba6a)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/routers/routes/web.go:107 (0x2dde7a4)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/vendor/github.com/go-chi/chi/middleware/strip.go:30 (0x2dd4927)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/vendor/github.com/chi-middleware/proxy/middleware.go:37 (0x2dcfe53)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/routers/routes/web.go:63 (0x2dde2e7)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /source/vendor/github.com/go-chi/chi/mux.go:87 (0x27492f0)
        /source/modules/web/route.go:298 (0x2ba151a)
        /source/vendor/github.com/gorilla/context/context.go:141 (0x1d980fa)
        /usr/local/go/src/net/http/server.go:2069 (0x13858aa)
        /usr/local/go/src/net/http/server.go:2887 (0x1388ee9)
        /usr/local/go/src/net/http/server.go:1952 (0x138436c)
        /usr/local/go/src/runtime/asm_amd64.s:1371 (0x1026640)

## Description

When I try to change any authentication source I get a 500 site and the above mentionesd runtime error and it is impossible to change any LDAP settings afterwards.

### PR DESCRIPTION
**Pull request by zeripath**
Unfortunately due a bug in xorm (see https://gitea.com/xorm/xorm/pulls/1957) updating
loginsources (and a few other places) on MSSQL causes them to become corrupted. (#16252)

Whilst waiting for the referenced PR to be merged and to handle the corrupted
loginsources correctly we need to add a wrapper to the `FromDB()` methods to look
for and ignore the misplaced BOMs that have been added.

Fix #16252

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: models/login_source.go [hunk 1]
```diff
@@ -70,15 +70,23 @@ var (
 	_ convert.Conversion = &SSPIConfig{}
 )
 
+func jsonUnmarshal(bs []byte, v interface{}) error {
+	json := jsoniter.ConfigCompatibleWithStandardLibrary
+	err := json.Unmarshal(bs, &v)
+	if err != nil && len(bs) > 2 && bs[0] == 0xff && bs[1] == 0xfe {
+		err = json.Unmarshal(bs, &v)
```
Comment (zeripath) [inline models/login_source.go]:
```suggestion
		err = json.Unmarshal(bs[2:], &v)
```

**Comment by zeripath** [comment]
ah crap this affects a lot more than just login_source

File: models/login_source.go [hunk 2]
```diff
@@ -70,15 +70,25 @@ var (
 	_ convert.Conversion = &SSPIConfig{}
 )
 
+// jsonUnmarshalIgnoreErroneousBOM - due to a bug in xorm (see https://gitea.com/xorm/xorm/pulls/1957) >
```
Comment (zeripath) [inline models/login_source.go]:
```suggestion
// jsonUnmarshalIgnoreErroneousBOM - due to a bug in xorm (see https://gitea.com/xorm/xorm/pulls/1957) - it's
```

**Review by noerw — COMMENTED** [review]
Is `Task.PayloadContent` not affected? That's also serializing to JSON

**Comment by zeripath** [comment]
The bug only affects `convert.Conversion` as far as I can see.

**Comment by zeripath** [comment]
`Task.PayloadContent` doesn't implement `convert.Conversion` so as far as Xorm is concerned it's just a string.

Take a look at the linked xorm PR.

**Review by noerw — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

File: models/login_source.go [hunk 3]
```diff
@@ -70,15 +70,25 @@ var (
 	_ convert.Conversion = &SSPIConfig{}
 )
 
+// jsonUnmarshalIgnoreErroneousBOM - due to a bug in xorm (see https://gitea.com/xorm/xorm/pulls/1957) - it's
+// possible that a Blob may gain an unwanted prefix of 0xff 0xfe.
+func jsonUnmarshalIgnoreErroneousBOM(bs []byte, v interface{}) error {
+	json := jsoniter.ConfigCompatibleWithStandardLibrary
+	err := json.Unmarshal(bs, &v)
```
Comment (zeripath) [inline models/login_source.go:77]: This is double indirection!

File: models/login_source.go [hunk 4]
```diff
@@ -70,15 +70,25 @@ var (
...
+	if err != nil && len(bs) > 2 && bs[0] == 0xff && bs[1] == 0xfe {
+		err = json.Unmarshal(bs[2:], &v)
```
Comment (zeripath) [inline models/login_source.go:79]: This is double indirection!

File: models/login_source.go (hunk 3 above)
Comment (zeripath) [reply models/login_source.go:77]:
```suggestion
	err := json.Unmarshal(bs, v)
```

File: models/login_source.go (hunk 4 above)
Comment (zeripath) [reply models/login_source.go:79]:
```suggestion
		err = json.Unmarshal(bs[2:], v)
```

**Pull request merged**